### PR TITLE
datanames slot in teal_transform_module

### LINKED ISSUE #1298: [Bug]: filter panel for related datanames of ANL are not shown
### What happened?

When `ANL` (or any new dataset) is created in `teal_transform_module` and `teal_module` is based on this **new** dataset only, then related datasets are ignored when building an application. For related datanames filter panel is not shown as they are not listed in `module$datanames` but only needed in `teal_transform_module` to successfully evaluate transform code.

We need a way to include related datanames also in the filter-panel and be able to filter them (in order to filter ANL indirectly).
Matter was discussed by the team and we sense that `teal_transform_module` needs `datanames` argument which can be set by app-developer, but also can have relevant default values.

<details>
<summary> example app </summary>
 
```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

# pkgload::load_all("teal.data")
pkgload::load_all("teal.slice")
pkgload::load_all("teal")

anl_transformer <- list(
  teal_transform_module(
    label = "ANL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL (ADSL + ADTTE)")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(), {
            ANL <- dplyr::inner_join(
              ADSL,
              ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
              by = c("USUBJID", "STUDYID")
            )
          })
        })
      })
    }
  )
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- teal_data() |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- head(teal.data::rADSL, n = n)
              ADTTE <- teal.data::rADTTE
              aaa <- iris
              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            n = as.numeric(input$obs)
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "aaa", "ADSL")
        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = list(
    example_module("anl from transform", dataname = "ANL", transformers = anl_transformer)
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    include_varnames = list(
      ADSL = c("SEX", "AGE")
    )
  )
)

runApp(app)


```

</details>

P.S.

When creating wrappers for example `transform_merge_module` it should be very easy to determine "related datanames" as they would have to be specified to construct the merge call. We think that problem of `datanames` exists only for generic `teal_transform_module` in situations when `module$datanames` don't contain related datanames (normally they do contain).

### PR DESCRIPTION
**Pull request by gogonzo**
Closes #1298 

- Added `datanames` slot to the `teal_transform_module`
- Added tests

Now it works like this:

1. filter panel displays (and thus filters) datasets which are in `module$datanames` and in `module-transformers$datanames`.
2. summary displays row counts for datanames specified in `module$datanames`, `module-transformers$datanames` and everything which have been created in a transformer (for example ANL)
3. teal_module$server receives always only `module$datanames`

With above rules, there are following cases possible:

1. app developer specifies `module$datanames = "ADSL"` and adds transform which modifies ADSL (doesn't need anything else). 
    - If `module-transform$datanames = NULL` then only `ADSL` will be displayed in the filter panel
    - if `module-transform$datanames = c("ADTTE", "whatever")` then "ADTTE" and "whatever" will also be displayed and filtered even if not used by either transform and teal_module. `teal_module` though will always receive only module$datanames (ADSL).  
2. app developer specifies `module$datanames = "ADSL"` and adds transform which modifies ADSL based on other datasets (does need other datasets to work)
    -   If `module-transform$datanames = NULL` then only `ADSL` will be displayed in the filter panel and only ADSL will be filtered. Transform will still work but other datasets won't be filtered
    -  If `module-transform$datanames = c("ADTTE", "whatever")` then both will be displayed, filtered and passed to transform. 
3. app developer specified `module$datanames = "ANL"` and specifies a transform which uses "ADSL" and "ADTTE" to make ANL.
    - if `module-transform$datanames = NULL` then nothing will be displayed in the filter panel as there is no information about "ADSL" and "ADTTE". Transform would still work and ANL would be created based on **unfiltered** ADSL and ADTTE
    - if `module-transform$datanames = c("ADSL", "ADTTE")` then everything is working as expected. ADSL and ADTTE is displayed in the filter panel and filtered are passed to transform module.
    
<details>
<summary>app example</summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

# pkgload::load_all("teal.data")
pkgload::load_all("teal")

trans <- list(
  anl_w_datanames = teal_transform_module(
    label = "ANL with datanames",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL based on specified ADSL, ADTTE"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = "ADTTE"
  ),
  anl_wout_datanames = teal_transform_module(
    label = "ANL without datanames",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL based on unspecified ADSL, ADTTE"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    }
  ),
  adsl_w_datanames = teal_transform_module(
    label = "modify ADSL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer modifies ADSL based on specified ADTTE")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ADTTE_summary <- ADTTE |>
                dplyr::group_by(STUDYID, USUBJID) |>
                dplyr::summarize(PARAMCD_AVAL = paste(paste(PARAMCD, "-", AVAL), collapse = "; "))
              ADSL <- dplyr::left_join(ADSL, ADTTE_summary)
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = c("ADSL", "ADTTE")
  ),
  adsl_wout_datanames = teal_transform_module(
    label = "modify ADSL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer modifies ADSL based on unspecified ADTTE")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ADTTE_summary <- ADTTE |>
                dplyr::group_by(STUDYID, USUBJID) |>
                dplyr::summarize(PARAMCD_AVAL = paste(sprintf("%s - %.2f", PARAMCD, AVAL), collapse = "; "))
              ADSL <- dplyr::left_join(ADSL, ADTTE_summary)
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = "ADSL"
  )
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- teal_data(a = NULL) |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- head(teal.data::rADSL, n = n)
              ADTTE <- teal.data::rADTTE
              iris <- iris
              iris_raw <- iris
              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            n = as.numeric(input$obs)
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "iris_raw", "iris", "ADSL")
        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = list(
    example_module("anl - transform w/ datanames", dataname = "ANL", transformers = trans["anl_w_datanames"]),
    example_module("anl - transform w/o datanames", dataname = "ANL", transformers = trans["anl_wout_datanames"]),
    example_module("adsl - transform w/ datanames", dataname = "ADSL", transformers = trans["adsl_w_datanames"]),
    example_module("adsl - transform w/o datanames", dataname = "ADSL", transformers = trans["adsl_wout_datanames"])
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    teal_slice("ADTTE", "PARAMCD", selected = "CRSD"),
    include_varnames = list(
      ADSL = c("SEX", "AGE")
    )
  )
)

runApp(app)

```
  
</details>

### CODE REVIEW

**Review by donyunardi — COMMENTED** [review]
This is a great start! 

I have a few comments:

### `module$datanames` not provided
When `modules$datanames` is not provided (i.e. `example_module("adsl - no transformer no datanames in module")`, the default value for `modules$datanames` is `all` which means the module has access to all datasets. 

However, I don’t see the `ANL` object and only datasets in `teal_data_module` available to the module. Shouldn't ANL be available in the module or does the user need to specify it explicitly in `module$datanames`?

### `teal_transform_module$datanames` vs `module$datanames`
From my understanding of the current design, when `teal_transform_module$datanames` is provided, it surfaces the datasets to the filter panel, allowing users to subset the datasets (and their parents if they have one), triggering the transformation to be rerun. 

> teal_module though will always receive only module$datanames (ADSL).

Ultimately, only the datasets specified in `module$datanames` are available to the module (and its parents if they have one). 

Since we’re using the same term `datanames`, I have a slight concern that this might cause confusion for users, who might think that `teal_transform_module$datanames` has the same scope or purpose as `module$datanames`. I may be overthinking this, but I’m curious to hear what everyone else thinks.

### Data Summary UI
The format of the Data Summary UI differs when the module only needs the new dataset (e.g., missing a bold header). Might need adjustment for consistency?

Only ANL


With other datasets

**Comment by gogonzo** [comment]
> When modules$datanames is not provided (i.e. example_module("adsl - no transformer no datanames in module"), the default value for modules$datanames is all which means the module has access to all datasets.
>
> However, I don’t see the ANL object and only datasets in teal_data_module available to the module. Shouldn't ANL be available in the module or does the user need to specify it explicitly in module$datanames

If `modules$datanames == "all"` then ANL will also be available in summary and in teal_module. "adsl - no transformer no datanames in module" doesn't create ANL. In (`"adsl - *"`) modules, ADSL is modified based on ADTTE.

> From my understanding of the current design, when teal_transform_module$datanames is provided, it surfaces the datasets to the filter panel, allowing users to subset the datasets (and their parents if they have one), triggering the transformation to be rerun.

Exactly.

> Ultimately, only the datasets specified in module$datanames are available to the module (and its parents if they have one).

Exactly.

> Since we’re using the same term datanames, I have a slight concern that this might cause confusion for users, who might think that teal_transform_module$datanames has the same scope or purpose as module$datanames. I may be overthinking this, but I’m curious to hear what everyone else thinks.

I agree, this is why I didn't just `@inheritParams module` in the documentation of `teal_transform_module`. I think this is a relevant concern and I would focus on the documentation and maybe parameter naming. I'll think more about `datanames` in general.

> The format of the Data Summary UI differs when the module only needs the new dataset (e.g., missing a bold header). Might need adjustment for consistency?

Good point. I'll fix it here 👍

**Comment by donyunardi** [comment]
> If modules$datanames == "all" then ANL will also be available in summary and in teal_module. "adsl - no transformer no datanames in module" doesn't create ANL. In ("adsl - *") modules, ADSL is modified based on ADTTE.

I don't see `ANL` available to the module when I don't provide the `datanames` argument or if I set `datanames = 'all'`

<details>

<summary> Example Code </summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("../teal.data")
pkgload::load_all("../teal")

# library(teal)

anl <- teal_transform_module(
  label = "ANL with datanames",
  ui = function(id) {
    ns <- NS(id)
    tagList(
      div("This transformer adds ANL based on specified ADSL, ADTTE"),
      numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        within(data(),
          {
            ANL <- dplyr::inner_join(
              head(ADSL, nobs),
              ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
              by = c("USUBJID", "STUDYID")
            )
          },
          nobs = input$obs
        )
      })
    })
  }
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- teal_data(a = NULL) |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- head(teal.data::rADSL, n = n)
              ADTTE <- teal.data::rADTTE
              iris <- iris
              iris_raw <- iris
              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            n = as.numeric(input$obs)
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "iris_raw", "iris", "ADSL")
        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = list(
    example_module("anl - no module$datanames provided", transformers = list(anl)),
    example_module("anl - module$datanames set to all", datanames = "all", transformers = list(anl))
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    include_varnames = list(
      ADSL = c("SEX", "AGE")
    )
  )
)

runApp(app)
```

</details>





@gogonzo Do you get a different result?

<details>

<summary>SessionInfo</summary>

```r
r$> sessionInfo()
R version 4.4.1 (2024-06-14)
Platform: x86_64-apple-darwin20
Running under: macOS Ventura 13.6.7

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Los_Angeles
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] teal_0.15.2.9056      teal.slice_0.5.1.9010 shiny_1.8.1.1         teal.data_0.6.0.9008  testthat_3.2.1.1      teal.code_0.5.0.9007 

loaded via a namespace (and not attached):
 [1] gtable_0.3.5             xfun_0.43                bslib_0.7.0              ggplot2_3.5.1            htmlwidgets_1.6.4        shinyjs_2.1.0           
 [7] teal.widgets_0.4.2.9017  vctrs_0.6.5              tools_4.4.1              generics_0.1.3           parallel_4.4.1           tibble_3.2.1            
[13] fansi_1.0.6              pkgconfig_2.0.3          data.table_1.15.4        checkmate_2.3.1          desc_1.4.3               lifecycle_1.0.4         
[19] compiler_4.4.1           brio_1.1.5               munsell_0.5.1            fontawesome_0.5.2        codetools_0.2-20         httpuv_1.6.15           
[25] shinyWidgets_0.8.6       htmltools_0.5.8.1        sass_0.4.9               lazyeval_0.2.2           yaml_2.3.8               plotly_4.10.4           
[31] tidyr_1.3.1              later_1.3.2              pillar_1.9.0             jquerylib_0.1.4          cachem_1.0.8             mime_0.12               
[37] parallelly_1.37.1        tidyselect_1.2.1         digest_0.6.35            future_1.33.2            purrr_1.0.2              dplyr_1.1.4             
[43] listenv_0.9.1            rprojroot_2.0.4          fastmap_1.1.1            grid_4.4.1               colorspace_2.1-0         cli_3.6.2               
[49] logger_0.3.0             magrittr_2.0.3           pkgbuild_1.4.4           utf8_1.2.4               withr_3.0.0              teal.reporter_0.3.1.9011
[55] scales_1.3.0             promises_1.3.0           backports_1.4.1          httr_1.4.7               rmarkdown_2.26           globals_0.16.3          
[61] memoise_2.0.1            evaluate_0.23            knitr_1.46               shinycssloaders_1.0.0    viridisLite_0.4.2        rlang_1.1.3             
[67] Rcpp_1.0.12              xtable_1.8-4             glue_1.7.0               formatR_1.14             renv_1.0.7               pkgload_1.3.4           
[73] jsonlite_1.8.8           teal.logger_0.2.0.9006   R6_2.5.1
```

</details>

**Comment by gogonzo** [comment]
> > If modules$datanames == "all" then ANL will also be available in summary and in teal_module. "adsl - no transformer no datanames in module" doesn't create ANL. In ("adsl - *") modules, ADSL is modified based on ADTTE.
> 
> I don't see `ANL` available to the module when I don't provide the `module$datanames` argument or if I set `module$datanames = 'all'`
> 

Yeah, thanks for this example. I have something to fix here 👍

**Comment by kumamiao** [comment]
>Since we’re using the same term datanames, I have a slight concern that this might cause confusion for users, who might think that teal_transform_module$datanames has the same scope or purpose as module$datanames. I may be overthinking this, but I’m curious to hear what everyone else thinks.

> I agree, this is why I didn't just @inheritParams module in the documentation of teal_transform_module. I think this is a relevant concern and I would focus on the documentation and maybe parameter naming. I'll think more about datanames in general.

Just wanted to echo with @donyunardi 's concern on the parameter name for `datanames` here, would be great to name it differently, thanks @gogonzo for trying to address this.

> @gogonzo Do you get a different result?

I get the same result as you do, I can only get `ANL` when I set `module$datanames` to `ANL`.

**Comment by gogonzo** [comment]
@kumamiao @donyunardi 
Dony's case reminded me what I forgot to mention. If `module@datanames == "all"` then module gets datanames specified in `teal_data` object. But app developer doesn't need to specifiy `datanames(data) <-` because if object doesn't contain any proxy variables then all bindings in `teal_data` are used. This means that the process of determining datanames for the module is multi-staged:
- `module$datanames` if specified
- else `datanames(data)` if specified
- else `ls(data@env)`

Same applies to the `module-transform$datanames` when determining datanames for transform. This gets little complicated and I'll spend some time on simplifying this.

**Comment by gogonzo** [comment]
@donyunardi @kumamiao 
The bug you've detected is marked with the red color on this diagram. This is __how datanames are determined for teal_module__  
image

Sidebar gets `datanames` based on the same logic but replace `module$datanames` with `union(module$datanames, module-transform$datanames)`. (because sidebar needs to receive datanames needed by module and needed by transform).

In my opinion this is getting too complicated:
1. I've raised already my opinion about `datanames = NULL` that it shouldn't be handled this way. I can't find the issue where we had the discussion about this
2. Do we really need the whole logic when `module$datanames = "all"`, maybe `datanames(data)` should be ignored and we should show all datasets (`ls(<teal_data>@env)`) when `module$datanames = "all"` or some when `module$datanames = <some datasets>`. This will require app developer always specify `module$datanames` when `@env` contains some temp variables created in the preprocessing.

**Comment by donyunardi** [comment]
> I've raised already my opinion about datanames = NULL that it shouldn't be handled this way. I can't find the issue where we had the discussion about this

We discussed it [here](https://github.com/insightsengineering/teal/issues/905).

> maybe datanames(data) should be ignored and we should show all datasets (ls(<teal_data>@env)) when module$datanames = "all" or some when module$datanames = <some datasets>. This will require app developer always specify module$datanames when @env contains some temp variables created in the preprocessing.

I’m open to this approach. Syntactically, it makes sense to me. If a user specifies `datanames = 'all'`, then teal will provide all the objects in `teal_data`. However, as you pointed out, this might include objects other than `data.frame`. Initially, I considered having teal filter these objects to show only `data.frame` objects, but I’ve reconsidered. It would be better for the module itself to handle this filtering. The teal module can be designed to display only `data.frame` objects (or other specific objects) as needed.

For now, I'm leaning toward just giving the user all objects.

If we decide to move forward with this behavior, we'll need to carefully assess how it will affect all the modules that currently have `datanames = 'all'` in both tmg and tmc.

### Idea

Have you considered adding a new argument to the transformer to specify additional data objects that should be appended to the `teal_data` object?

Code can look something like this:
```r
teal_transform_module <- function(label, ui, server, datanames, data_to_be_added_to_teal_data)
```

User can define the value(s) of the `new_data_object` to specify which objects they want to pull from the transform module. Behind the scene, we will then append this information: `datanames(data) <- c(datanames(data), data_to_be_added_to_teal_data)`.

**Comment by gogonzo** [comment]
Let me rephrase once again my thoughts. We need to know few things:

1. `module$datanames` defines what teal_module needs.
2. We need a way to specify teal_transform_module needs - this is a subject of this PR and I enabled it though `transform$dataname` 
3. I believe that we can't just base on `module$datanames` as `teal_transform_module` might need completely different datasets in order to create/modify datasets needed by the `teal_module`
4. `teal_transform_module` might create a new dataset which should be included in the teal_module if its name matches `module$datanames` - matches either by being an element of `module$datanames` or when `module$datanames == "all"`

I think that above points define exactly what we need and we can discuss how provide simple logic for app-developer and not overcomplicate internal code. In this PR I've added `$datanames` to the `teal_transform_module` to optionally specify what datasets are needed for `teal_transform_module` to perform. Since we pass data in following order: `filter-panel -> teal_transform_module -> teal_module`, we need `filter-panel` to display and filter `datanames = c(transform$datanames, module$datanames)`. This is why I think (1) (2) is unavoidable and shouldn't be discussed.  

------

### Issue: `module$datanames == "all"` and new datanames added in transform.

The real problem is what @donyunardi [noticed](https://github.com/insightsengineering/teal/pull/1327#issuecomment-2305219010) that if `module$datanames == "all"` and if `teal_data@datanames` is specified, then dataset is ignored (because it is not specified in the teal_data object. This is the logic from the moment when @donyunardi made a comment:

image

Given above, I've made a fix and introduced an extra step in teal internals to figure out what is needed (automatically), let's call it a **solution 1 - determine datanames automatically**:

image

Then I realised that this gets pretty complicated and provoked discussion, by illustrating **solution 2 - simply determine datanames from the teal_data environment**

image

>Do we really need the whole logic when module$datanames = "all", maybe datanames(data) should be ignored and we should show all datasets (ls(<teal_data>@env)) when module$datanames = "all" or some when module$datanames = <some datasets>. This will require app developer always specify module$datanames when @env contains some temp variables created in the preprocessing. Diagram would look like this

In Today's meeting @pawelru proposed a **solution 3 - app developer needs to add extra items to @datanames**, but IMO forcing app-developer to append datanames only when `data@datanames` is specified and something new is added is to specific condition, which would result that app-developer might not remember what to specify when (currently needs to understand relation between module$datanames, data@datanames)

image

So I said, instead of forcing app-developer to update datanames in specific scenario, we should force it always, so that @datanames is always up to date **scenario 4 - allways specify @datanames**

image


--- 

I personally like (4) and (2). I would like (2) the most if we find a simple way to exclude proxy datanames not needed for teal_module.

**Comment by kumamiao** [comment]
Hi @gogonzo , thank you so much for the quick and comprehensive summary!!

Like you, I like (2) the most if we find a good way to only get needed data for `teal_module`, but otherwise I do not have strong preferences on either (2) or (4), both would make sense. To me personally, I feel (2) gives more flexibility to the user - if I'm understanding this, if user does not like all data@env objects to be displayed, user can choose to do what they need to do in option (4)?

I tested the current PR and things are working well, but could you please check if listed below are the expected behavior under option (2)?
a. In the last 2 modules, I specified `module$datanames`, and the subset of dataframes are valid on the left encoding panel, but the Active Filter Summary still displays both subset of dataframe + all newly added env objects
b. newly created data in transform are not listed under Active Filter Variables?

<details>
<summary>app example</summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("../teal.data")
pkgload::load_all("../teal")

anl <- teal_transform_module(
  label = "ANL with datanames",
  ui = function(id) {
    ns <- NS(id)
    tagList(
      div("This transformer adds ANL based on specified ADSL, ADTTE"),
      numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        within(data(),
               {
                 sub_iris <- head(iris)
                 ANL <- dplyr::inner_join(
                   head(ADSL, nobs),
                   ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                   by = c("USUBJID", "STUDYID")
                 )
                 tmp_obj1 <- 100
                 tmp_obj2 <- "test"
               },
               nobs = input$obs
        )
      })
    })
  }
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- teal_data(a = NULL) |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- head(teal.data::rADSL, n = n)
              ADTTE <- teal.data::rADTTE
              iris <- iris
              iris_raw <- iris
              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            n = as.numeric(input$obs)
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "iris_raw", "iris", "ADSL")
        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = list(
    example_module("NULL", datanames = NULL, transformers = list(anl)),
    example_module("no call", transformers = list(anl)),
    example_module("all", datanames = "all", transformers = list(anl)),
    example_module("ANL", datanames = "ANL", transformers = list(anl)),
    example_module("ANL+iris", datanames = c("ANL", "iris"), transformers = list(anl))
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    include_varnames = list(
      ADSL = c("SEX", "AGE")
    )
  )
)

runApp(app)

```
  
</details>

**Comment by gogonzo** [comment]
> In the last 2 modules, I specified module$datanames, and the subset of dataframes are valid on the left encoding panel, but the Active Filter Summary still displays both subset of dataframe + all newly added env objects

Good point @kumamiao I need to fix it

**Comment by donyunardi** [comment]
Thanks for the laying down all the conditions nicely @gogonzo. 

Do you mind helping me trying to understand this statement?

> app developer needs to add to data@datanames when module$datanames == "all" and when transform adds some new dataset which has to be passed to module.

When/where would user add/update `data@datanames`? Would it be in the server of teal_transform_module? 
if so, then if you look at @kumamiao [example code](https://github.com/insightsengineering/teal/pull/1327#issuecomment-2311753572), she has this line:
```
example_module("all", datanames = "all", transformers = list(anl)),
```

So here,`module$datanames` is `all` and there is no statement to add `ANL` object in `data$datanames` inside the  server of `anl` teal_transform_module. However, I see `ANL` is being passed to the module.



From what I observed, it doesn't look like app developer needs to add/update `data$datanames` when `module$datanames == "all"`.

**Comment by gogonzo** [comment]
> From what I observed, it doesn't look like app developer needs to add/update data$datanames when module$datanames == "all"

Sorry @donyunardi my bad - I didn't remove lines which were always adding everything from `teal_transform_module` - now it works as described.

**Comment by gogonzo** [comment]
Closing this PR in favour of https://github.com/insightsengineering/teal/pull/1334
